mep-0001-ray-future-based-shuffle.md

```diff
@@ -0,0 +1,161 @@
+```
+When running on ray, `source_keys`/`source_idxes`/`source_mappers` will be set to None to reduce meta overhead.
+- Make shuffle operands mapper produce output with deterministic number and order.
+   - `TensorReshape` and `TensorBinCount` mapper needs change.
+- _bagging.py uses `source_idxes`, which also needs to be refactored.
+ 
+#### Graph Changes
+Add an option to replace `ShuffleProxy` to `FetchShuffleByIndex` when building subtask graph.
+ 
+#### Using a ShuffleManager to manage Ray shuffle execution
+`ShuffleManager` is defined in `mars/services/task/execution/ray/shuffle.py` and only used by Ray executor. When `TaskProcessorActor` invokes `RayTaskExecutor.execute_subtask_graph` to execute a subtask graph, The `RayTaskExecutor` will create a `ShuffleManager` to manage shuffle execution.
+ 
+The `ShuffleManager` will do following things:
+- Build mapper index which is a dict from mapper subtask to (`shuffle_index`, `mapper_index`). A subtask graph may have multiple groups of shuffles, `shuffle_index` indicates which shuffle a mapper subtask belongs to. `mapper_index` is the ordinal in all mapper subtasks of current shuffle.
+- Build reducer index which is a dict from reducer subtask to (`shuffle_index`, `reducer_ordinal`). A subtask graph may have multiple groups of shuffles, `shuffle_index` indicates which shuffle a reducer subtask belongs to. `reducer_ordinal` is the ordinal in all reducer subtasks of current shuffle. If some reducers are missing, `ShuffleManager` will fill `None` for those reducers index.
+- Recording mapper output object refs for all mapper subtasks, which will be used by later reducers to get all mapper inputs based on reducer ordinal.
+- Return `n_reducers` for a shuffle when passing a mapper/reducer subtask.
+- In the future, it will manage push-based shuffle scheduling and execution.
+ 
+The code skeleton is as follows:
+ 
+```python
+class ShuffleManager:
+    """Manage shuffle execution for ray by resolve dependencies between mappers outputs and reducers inputs based on
+    mapper and reducer index.
+    """
+
+    def __init__(self, subtask_graph):
+        self.subtask_graph = subtask_graph
+        # Build mapper index
+        # subtask: (shuffle_index, mapper_index)
+        # Build reducer index
+        # subtask: (shuffle_index, reducer_ordinal)
+
+    def has_shuffle(self) -> bool:
+        """
+        Returns
+        -------
+        bool
+            Whether current subtask graph has shuffle to execute
+        """
+        return self.num_shuffles > 0
+
+    def add_mapper_output_refs(self, subtask, output_object_refs: List["ray.ObjectRef"]):
+        """
+        Record mapper output ObjectRefs which will be used by reducers later.
+
+        Parameters
+        ----------
+        subtask
+        output_object_refs : List["ray.ObjectRef"]
+            Mapper output ObjectRefs.
+        """
+        pass
+
+    def get_reducer_input_refs(self, subtask: Subtask) -> List["ray.ObjectRef"]:
```

Comment:
Maybe we can move the `get_reducer_input_refs` and `get_n_reducers` to the `SubtaskGraph` in the future, then Ray backend can share the same logic with Mars backend.

Reply:
Actually, those refs are Ray ObjectRefs, and in future there will be more ray-specific Ray Object management  code when push-based shuffle is implemented for ray. I think it's not easy to share this logic with Mars backend

Reply:
I know the refs are Ray object refs. For Mars, they are the map data keys. Mars can use this logic to calculate the map data keys.

Reply:
When push-based shuffle are implemented, parts of those refs are mapper ouputs, parts of it are merge outputs. I think those logic shouldn't be put into SubtaskGraph